### PR TITLE
Effective VPA api

### PR DESCRIPTION
**Pull request by qmhu**

### CODE REVIEW

File: autoscaling/v1alpha1/evpa_types.go [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+	// considered while scaling up or scaling down.
+	// +optional
+	// +kubebuilder:validation:Type=integer
+	// +kubebuilder:default=3600
+	StabilizationWindowSeconds *int32 `json:"stabilizationWindowSeconds,omitempty"`
+}
+
+type ResourceName string
+
+type ResourceMetricList map[ResourceName]ResourceMetric
```
Comment (kitianFresh) [inline autoscaling/v1alpha1/evpa_types.go:111]: the name is a List, but it is a map in fact, it is a little misleading

File: autoscaling/v1alpha1/evpa_types.go [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+
+type ResourceMetric struct {
+	// averageUtilization is the target value of the average of the
+	// resource metric across all relevant pods.
+	// +optional
+	AverageUtilization *int32 `json:"averageUtilization,omitempty"`
```
Comment (kitianFresh) [inline autoscaling/v1alpha1/evpa_types.go:117]: the ResourceMetric name is misleading, how about a Value?

File: autoscaling/v1alpha1/evpa_types.go (hunk 1 above)
Comment (qmhu) [reply autoscaling/v1alpha1/evpa_types.go:111]: It's inspire by `v1.ResourceList`.

File: autoscaling/v1alpha1/evpa_types.go (hunk 2 above)
Comment (qmhu) [reply autoscaling/v1alpha1/evpa_types.go:117]: the `MetricTarget` from hpa have no `Value` in name and the `CustomMetricValue` in Metric API have the `Value` in name, so i'm ok for both, WDYT @mfanjie

File: autoscaling/v1alpha1/evpa_types.go [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+}
+
+// ContainerScalingPolicy define effective policy for vertical scaling in certain direction: Up or Down
+type ContainerScalingPolicy struct {
+	// ScaleMode controls Whether autoscaler is enabled for the container.
+	// The default is "Auto".
+	// +kubebuilder:validation:Type=string
+	// +kubebuilder:validation:Enum=Auto;Off
+	// +kubebuilder:default=Auto
+	ScaleMode *vpatypes.ContainerScalingMode `json:"mode,omitempty"`
```
Comment (mfanjie) [inline autoscaling/v1alpha1/evpa_types.go:95]: how the VPA initial mode is supported in EVPA?

File: autoscaling/v1alpha1/evpa_types.go (hunk 2 above)
Comment (mfanjie) [reply autoscaling/v1alpha1/evpa_types.go:117]: I vote for current ResourceMetric, coz in some API, Value(like 5) and Utilization like (5%) are different type of metrics.

File: autoscaling/v1alpha1/evpa_types.go [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+	// +optional
+	Type string `json:"type,omitempty"`
+
+	// Priority defines the priority for this estimator.
+	// +optional
+	Priority int `json:"priority,omitempty"`
+
+	// Config contains key value pairs for this estimator
+	// +optional
+	Config map[string]string `json:"config,omitempty"`
```
Comment (mfanjie) [inline autoscaling/v1alpha1/evpa_types.go:132]: can you add a bit more information about how the Config should be defined?

File: autoscaling/v1alpha1/evpa_types.go (hunk 2 above)
Comment (qmhu) [reply autoscaling/v1alpha1/evpa_types.go:117]: Nice reason, let's go with ResourceMetric

File: autoscaling/v1alpha1/evpa_types.go (hunk 3 above)
Comment (qmhu) [reply autoscaling/v1alpha1/evpa_types.go:95]: I think you are referring to `UpdatePolicy`. 

For initial mode, we should implement that with an admission controller.

**Pull request merged**